xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp

```diff
@@ -1715,6 +1715,9 @@ CDemuxStream* CDVDDemuxFFmpeg::AddStream(int streamIdx)
         if (!stereoMode.empty())
           st->stereo_mode = stereoMode;
 
+        side_data = av_stream_get_side_data(pStream, AV_PKT_DATA_DOVI_CONF, &size);
+        if (size > 0 && side_data)
+          pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'v', 'C');
```

Comment:
This code overwrites `codec_tag` unconditionally for all DV videos in any container. This is not correct.

i.e. .mp4 files which already know codec_tag must not be altered here. 

Also `av_stream_get_side_data` is not necessary (if is not used to parse side data) as `hdr_type` variable is already available.

These are my points although I don't know exact differences and details about different DV codec_tags. Based on this I think is more suitable something like this:

```c++
if (st->hdr_type == StreamHdrType::HDR_TYPE_DOLBYVISION)
{
  if (m_bMatroska)
  {
    pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'v', 'C');
  }
  else if (pStream->codecpar->codec_id == AV_CODEC_ID_HEVC &&
           pStream->codecpar->codec_tag == 0)
  {
    pStream->codecpar->codec_tag = MKTAG('d', 'v', 'h', 'e');
  }
}
```

Reply:
hey @thexai, thanks for giving an input here. 

I started with a matroska check but figure the container would not make any difference and also since every Dolby Vision tag results in the same <code>m_mime</code> and <code>m_format</code> being passed to the decoder, it would actually be a moot point. 

In the above code we are adding the MKBETAG to every matroska file and looking here (https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/matroskaenc.c#L1592), following your idea of preserving what exists but fixing what is missing:

<code>
  if (st->hdr_type == StreamHdrType::HDR_TYPE_DOLBYVISION)
  {
    if (m_bMatroska && pStream->codecpar->codec_tag == 0)
    {
      side_data = av_stream_get_side_data(pStream, AV_PKT_DATA_DOVI_CONF, &size);
      if (size > 0 && side_data)
      {
        auto dovi = reinterpret_cast<AVDOVIDecoderConfigurationRecord*>(side_data);
        if (dovi->dv_profile > 7)
          pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'v', 'C');
        else
          pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'c', 'C');
      }
    }
    else if (pStream->codecpar->codec_id == AV_CODEC_ID_HEVC && pStream->codecpar->codec_tag == 0)
    {
      pStream->codecpar->codec_tag = MKTAG('d', 'v', 'h', 'e');
    }
  }
</code>

This would follow exactly the standard used in FFMpeg and will only fill the blanks where it's needed at the expense of having to check the side_data. That was the original idea in the PR. 

If this is the case we would have to check the tags:

<code>
bool isDvh1 = (m_hints.codec_tag == MKTAG('d', 'v', 'h', '1') ||
                            m_hints.codec_tag == MKBETAG('d', 'v', 'v', 'C') ||
                             m_hints.codec_tag == MKBETAG('d', 'v', 'c', 'C'));
</code>

Let me know what you think.

Reply:
There are DV files that have codec_tag != 0,  for example 36. Those will not be detected as DV with above code. Existing way to transfer the information from Demuxer to Codec is to use hints. Please consider changes from https://github.com/xbmc/xbmc/compare/master...sscobici:xbmc:dv-android or let me know if I need to raise a separate PR.

Reply:
@fandangos 

As `codec_tag == 0` condition is common can be simplified a bit:

```c++
        // Add DoVi codec_tag if is missing
        if (st->hdr_type == StreamHdrType::HDR_TYPE_DOLBYVISION &&
            pStream->codecpar->codec_tag == 0)
        {
          if (m_bMatroska)
          {
            side_data = av_stream_get_side_data(pStream, AV_PKT_DATA_DOVI_CONF, &size);
            if (size > 0 && side_data)
            {
              auto dovi = reinterpret_cast<AVDOVIDecoderConfigurationRecord*>(side_data);
              if (dovi->dv_profile > 7)
                pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'v', 'C');
              else
                pStream->codecpar->codec_tag = MKBETAG('d', 'v', 'c', 'C');
            }
          }
          else if (pStream->codecpar->codec_id == AV_CODEC_ID_HEVC)
          {
            pStream->codecpar->codec_tag = MKTAG('d', 'v', 'h', 'e');
          }
        }
```

@sscobici 

Demuxer may add/guess missing information but should not overwrite/change/ignore already valid infos.

> There are DV files that have codec_tag != 0.

1. Is this code_tag a valid DV code_tag?
2. Is this file really DV playable file?

If 1 and 2 are true then this already valid `codec_tag` not should be overwrite but be added to `DVDVideoCodecAndroidMediaCodec`

Another important point is that demuxer code is common for all platforms and "Add DoVi codec_tag if is missing" seems a task that can be done for all platforms (although now only android will use it).

Reply:
@thexai  Here is an DV mp4 file that have codec_tag hev1 which is currently played as HDR.
I'm not sure that the code you provided will fix DV detection.

http://media.developer.dolby.com/DolbyVision_Atmos/mp4/P81_GlassBlowing2_3840x2160%4059.94fps_15200kbps_fmp4.mp4
HDR format : Dolby Vision, Version 1.0, dvhe.08.09, BL+RPU, HDR10 compatible / SMPTE ST 2086, HDR10 compatible
Codec ID : hev1

Reply:
> If 1 and 2 are true then this already valid codec_tag not should be overwrite or ignore but be added to DVDVideoCodecAndroidMediaCodec

Then add `MKTAG('h', 'e', 'v', '1')` to  DVDVideoCodecAndroidMediaCodec

Reply:
Adding <code>MKTAG('h', 'e', 'v', '1')</code> to DVDVideoCodecAndroidMediaCodec will cause some HDR10 to used the Dolby Vision rendered. 

The code @sscobici provide in his git checks against <code>StreamHdrType::HDR_TYPE_DOLBYVISION</code> and this can be done here with <code>st->hdr_type == StreamHdrType::HDR_TYPE_DOLBYVISION</code>

My concern with the code above is that it will set <code>MKTAG('d', 'v', 'h', 'e');</code> to all videos with CODEC ID HEVC. 

I believe a better <code>else if</code> would be something like this:
<code>
else if (st->hdr_type == StreamHdrType::HDR_TYPE_DOLBYVISION && pStream->codecpar->codec_id == AV_CODEC_ID_HEVC)
</code>

Reply:
I'm not sure I'm following you, because in my proposal code I don't need that.
I will create another PR as alternative approach of fixing DV detection for Android where people will provide the feedback.

Reply:
Honestly, that's best.

Reply:
Scratch the above about <code>MKTAG('h', 'e', 'v', '1')</code>.

The code checks if the display anbd the mediaCodec supports Dolby Vision. 
So it's perfectly fine to add <code>MKTAG('h', 'e', 'v', '1')</code> to DVDVideoCodecAndroidMediaCodec.

Also this will not force Dolby Vision inconditionally for everything that hints as <code>StreamHdrType::HDR_TYPE_DOLBYVISION</code> and also will provide differentiation for different Dolby Vision type files, if the need to set different codecs and renderers per tag in the future.

I'm compiling and I'm going to test against the samples. 
If it's good, I'll update this PR, provide the compiled apk for testing and we should be good to go.

Reply:
I have added both suggestions to this PR giving credits to @sscobici. 

So this can be an option that both fixes <code>codec_tag</code> when it's missing following FFMpeg libavformat. 

As another PR is open (https://github.com/xbmc/xbmc/pull/22423), feel free to use any.

One caveat of using the provided method is that it will force Dolby Vision if side_data is present by is invalid or has a value of zero.

But I don't see any way to use the Dolby Vision decoder for files without the <code>codec_tag</code>, without actually patching and overwritting the <code>codec_tag</code>.

So, feel free to use any as I see for now @sscobici is a more universal and better approach.